Upgrade from profile2

bugfolder opened the pull request:
This PR responds to this issue: https://github.com/backdrop-contrib/profile/issues/5

It adds a hook_update_N() implementation that migrates any existing Profile2 profiles into this module and cleans up the db and config directory.

It also adds some discussion of the migration to the module README file.

And it addresses an issue with the update numbering. The original first update was numbered 10000, rather than 1000. This change corrects the hook_update_N numbering and fixes the numbering in the system table. Existing installations will still work, with the fix applied as part of the current upgrade process.

bugfolder committed:
Correct update numbering and migrate Profile2 data

Correct the update numbering so that it starts from 1000, not 10000. Add an update to migrate existing D7 Profile2 data into this module.

bugfolder committed:
Expand README

Add more material to the README.md file, including discussion about upgrading from Drupal. Add links to README in various places.

bugfolder committed:
Simplify cleanup in hook_requirements

Simplify the cleanup in hook_requirements by using installed_schema functions. (H/T @Laryn.)

bugfolder committed:
Replace undefined constant

PROFILE_STORAGE_NORMAL isn’t defined for the .install file; use explicit value.

herbdool commented on README.md:
```diff
@@ -3,31 +3,65 @@ Profile
+Documentation
+-------------
+
+Additional documentation is located in the [Wiki](https://github.com/backdrop-contrib/me/wiki/Documentation).
```
Shouldn't this URL be pointing to "profile" and not "me"?

bugfolder replied:
Yes, fixed.

herbdool commented on README.md:
```diff
@@ -3,31 +3,65 @@ Profile
+Issues
+------
+
+Bugs and feature requests should be reported in the [Issue Queue](https://github.com/backdrop-contrib/me/issues).
```
Same here.

bugfolder replied:
Also fixed.

herbdool commented on README.md:
```diff
@@ -3,31 +3,65 @@ Profile
+Current Maintainers
+-------------------
+
+* [docwilmot](https://github.com/docwilmot)
```
Did @docwilmot agree to be the maintainer? If not we can leave this as "Seeking maintainers". Or add your name here @bugfolder?

bugfolder replied:
Ah, I'd elided the distinction between "porter" and "maintainer".

Though I've still got reservations about becoming the maintainer, it's better to have someone to respond, however limited their skills; so I've made myself the maintainer.

herbdool replied:
@bugfolder I think you'll do fine. If you've got any questions feel free to ask anytime. I was hesitant to maintain at first as well.

herbdool commented on profile.install:
```diff
@@ -86,9 +88,147 @@ function profile_uninstall() {
+ * proceed properly going forward.
+ */
+function profile_requirements($phase) {
+  if (backdrop_get_installed_schema_version('profile') == 10000) {
```
It's probably better to only have this check during the `update` and `install` phases. If that still works for this. Otherwise it'll always be running this query on every page load.

see https://api.backdropcms.org/api/backdrop/core%21modules%21system%21system.api.php/function/hook_requirements/1

bugfolder replied:
This is difficult. We'd had [a Zulip conversation about this issue](https://backdrop.zulipchat.com/#narrow/stream/218635-Backdrop/topic/hook_update_N.20numbering). The problem is that an earlier version gave the wrong update number (10000, not 1000), which means that any further properly numbered updates wouldn't get called (1001 being less than 10000), and use of numbers >7000 are not allowed anyhow. The issue isn't relevant for fresh installations.

[Laryn suggested using hook_requirements](https://backdrop.zulipchat.com/#narrow/stream/218635-Backdrop/topic/hook_update_N.20numbering/near/214141433), so I'd adopted that approach.

It looks like hook_requirements() is not called on every page load, but only during `install`, `update`, and called with `runtime` during display of the status page. Is that incorrect?

herbdool replied:
Yeah, I had seen that conversation. The idea might be the best option. I figured it might be possible to reduce the times when it checks the schema version by limiting it to only being during the install or update phases.

Just to be safe, you may also want to ensure that you return what the hook expects. So if you've got no actual status message to return, then do `return NULL` like here https://api.backdropcms.org/api/backdrop/core%21includes%21install.core.inc/function/install_verify_requirements/1.

bugfolder replied:
Even though NULL appears OK as a return value, the docs do say to return an array, so I've added an empty array as the return value.

Also removed the check during 'install' phase (since it's not needed at first installation).

And, upon further investigation, doing the check-and-fix during 'update' phase happens early enough (before checking update numbering) that the numbering will get fixed during the routine update.php that people are supposed to always run after upgrading a module.

If a person forgets to run update.php, the admin bar or status page normally warns them that there are unapplied updates; but with the mis-numbering, this won't happen with this module. However, the first time they check the status page, the 'runtime' phase will be called, which will fix the update numbering, and thereafter they'll get nagged to run update.php.

And last, it appears that hook_requirements() is not called on every page load, so there's no each-page penalty.

So, I think the current version now fully works!

herbdool reviewed (changes requested):
I've got just minor change requests. Overall looks pretty robust. I haven't tested it at all but the field conversion looks good to me. Thanks for your work on this!

As part of the Bug Fix Squad I could merge this PR - while it's a grey area that this is a bug, since there's no maintainer currently for this project, it's probably okay if get this module fixed.

bugfolder committed:
Fixed README.md

Fixed some errors in the README. Added self as maintainer.

bugfolder committed:
Limit db check to update/install in hook_requirements().

Don’t need to do the check at installation. Also return an empty array for all calls.

herbdool commented:
@bugfolder looks good now. I'll merge it once permissions are sorted out on this project.
